vm_phys.c

```diff
@@ -0,0 +1,249 @@
+  page_buddy->size = ORDER_SIZE(page_buddy->order) * PAGE_SIZE;
+
+  TAILQ_INSERT_HEAD(&seg->free_queues[order - 1], page, freeq);
+  TAILQ_INSERT_HEAD(&seg->free_queues[order - 1], page_buddy, freeq);
+  page_buddy->flags |= VM_FREE;
+}
+
+
+static void vm_phys_reserve_from_seg(vm_phys_seg_t *seg, vm_paddr_t start,
+                                     vm_paddr_t end) {
+  /* Split segment into smaller segments,
+   * remove them from free queue, mark space as reserved */
+  for (int i = VM_NFREEORDER - 1; i >= 0; i--) {
+    vm_page_t *pg_ptr;
+    vm_page_t *pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+    while (pg_it) {
+      if (pg_it->phys_addr >= start &&
+          pg_it->phys_addr + pg_it->size <= end) {
+        pg_ptr = pg_it;
+        while (pg_ptr++ < pg_it + ORDER_SIZE(pg_it->order))
+          pg_ptr->flags |= VM_RESERVED;
+        TAILQ_REMOVE(&seg->free_queues[i], pg_it, freeq);
+        /* Go to start of list because it's corrupted */
+        pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+        continue;
+      } else if (  (pg_it->phys_addr + pg_it->size > start &&
+                    pg_it->phys_addr < start) ||
+                   (pg_it->phys_addr + pg_it->size > end &&
+                    pg_it->phys_addr < end)) {
+        split_page(seg, pg_it);
+        /* Go to start of list because it's corrupted */
+        pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+        continue;
+      }
+      pg_it = TAILQ_NEXT(pg_it, freeq);
+    }
+  }
+}
+
+void vm_phys_reserve(vm_paddr_t start, vm_paddr_t end) {
+
+  vm_phys_seg_t *seg_it;
+  TAILQ_FOREACH(seg_it, &seglist, segq) {
+    if (seg_it->start <= start && seg_it->end >= end) {
+      vm_phys_reserve_from_seg(seg_it, start, end);
+      return;
+    }
+  }
+
+}
+
+static vm_page_t *vm_phys_alloc_from_seg(vm_phys_seg_t *seg, size_t order) {
+  vm_page_t *page = NULL;
+  size_t ord = order;
+  /* lowest non-empty order higer than order */
+  while (TAILQ_EMPTY(&seg->free_queues[ord]) && ord < VM_NFREEORDER)
+    ord++;
+
+  if (ord == VM_NFREEORDER)
+    return NULL;
+
+  while (TAILQ_EMPTY(&seg->free_queues[order])) {
+    split_page(seg, TAILQ_FIRST(&seg->free_queues[ord]));
+    ord--;
+  }
+
+  page = TAILQ_FIRST(&seg->free_queues[order]);
+  TAILQ_REMOVE(&seg->free_queues[order], page, freeq);
+
+  page->size = ORDER_SIZE(page->order) * PAGE_SIZE;
+  page->flags &= ~VM_FREE;
+  return page;
+}
+
+vm_page_t *vm_phys_alloc(size_t order) {
+
+  vm_phys_seg_t *seg_it;
+  TAILQ_FOREACH(seg_it, &seglist, segq) {
+    vm_page_t *res;
+    res = vm_phys_alloc_from_seg(seg_it, order);
+    if (res != NULL) return res;
+  }
+
+  return NULL;
+}
+
+static void vm_phys_free_from_seg(vm_phys_seg_t *seg, vm_page_t *page) {
+
+  page->flags |= VM_FREE;
+  /* check if page is reserved */
+  if ((page->flags & VM_RESERVED) != 0)
+    panic("vm_phys_free: trying to free reserved page: %p\n", page->phys_addr);
+
+  /* check if page was free'd before */
+  if ((page->flags & VM_FREE) != 0)
```

Comment:
You're good one, too bad i'm quite tired now and i just want to have it done, that's why i don't see such simple bugs.